File: packages/php/src/Inillucent.php
<?php

declare(strict_types=1);

namespace Inillucent;

/**
 * inillucent, from PHP.
 *
 * This drives the `inillucent` binary and returns its results as arrays. Each
 * call is a process, which makes it the right tool for the handful of calls a
 * migration, a console command or an agent harness makes, and the wrong tool
 * for a loop over a million rows.
 *
 * It is not an in-process driver, and saying so plainly is better than shipping
 * one that has never been run. A real binding goes through the C ABI in
 * `include/inillucent_driver.h` with `ext-ffi`, and `drivers/README.md` is
 * written to be followed by somebody doing exactly that - it names the five
 * rules a binding has to get right and why each one matters.
 *
 * ```php
 * $db = new Inillucent('app.rdb');
 * $db->exec('CREATE TABLE notes (id INTEGER PRIMARY KEY, body TEXT)');
 * $db->exec('INSERT INTO notes (body) VALUES (?1)', ['hello']);
 * foreach ($db->query('SELECT * FROM notes') as $row) {
 *     echo $row['id'], ' ', $row['body'], PHP_EOL;
 * }
 * ```
 */
final class Inillucent
{
    /** The four programs a release ships. */
    public const PROGRAMS = ['inillucent', 'inillucent-shell', 'inillucent-mcp', 'inillucent-migrate'];

    /**
     * @param string $path the database file, or ':memory:' for a scratch one
     * @param string|null $binary the inillucent program to run; found automatically when null
     * @param bool $readOnly refuse every statement that changes something
     * @param string|null $root refuse every path outside this directory
     */
    public function __construct(
        private string $path,
        private ?string $binary = null,
        private bool $readOnly = false,
        private ?string $root = null,
    ) {
        $this->binary ??= Locator::find();
    }

    /**
     * Encodes the bound values as the JSON text the command line reads.
     *
     * **`json_encode` cannot carry three things a parameter can be
     * (task-1979, D13, D15 and D16).** A binary string is not valid UTF-8, so
     * encoding one threw and PHP could not bind a BLOB at all; a NAN or an INF
     * throws too, where the honest answer is to name the value; and a negative
     * zero is written `-0`, which reads back as a positive zero.
     *
     * What is written instead: bytes become `{"blob":"<hex>"}`, a non-finite
     * number is refused here with the value in the message, and a negative zero
     * is written `-0.0`, which JSON's own grammar carries.
     *
     * @param list<mixed> $values the bound values, in order
     * @throws Error when a value has no SQL spelling
     */
    private static function encodeParams(array $values): string
    {
        $parts = [];
        foreach ($values as $value) {
            $parts[] = self::encodeParam($value);
        }

        return '[' . implode(',', $parts) . ']';
    }

    /**
     * Encodes one bound value. See {@see self::encodeParams}.
     *
     * @param mixed $value one element of the parameter list
     * @throws Error when the value has no SQL spelling
     */
    private static function encodeParam($value): string
    {
        if (is_float($value)) {
            if (is_nan($value) || is_infinite($value)) {
                throw new Error(
                    'a parameter cannot be NAN or INF: SQL has no spelling for either.',
                    'invalid_state'
                );
            }
            // `1 / $value` is a DivisionByZeroError in PHP 8, so the sign bit
            // is read off the bytes instead: `pack('e', ...)` is little-endian
            // IEEE 754, and the top bit of its last byte is the sign.
            if ($value === 0.0 && (ord(pack('e', $value)[7]) & 0x80) !== 0) {
                return '-0.0';
            }

            return json_encode($value, JSON_THROW_ON_ERROR | JSON_PRESERVE_ZERO_FRACTION);
        }
        // `preg_match('//u', ...)` rather than `mb_check_encoding`, because
        // mbstring is an extension a PHP build may not have and PCRE is not.
        if (is_string($value) && preg_match('//u', $value) !== 1) {
            return json_encode(['blob' => bin2hex($value)], JSON_THROW_ON_ERROR);
        }
        if (is_array($value)) {
            return self::encodeParams(array_values($value));
        }

        return json_encode($value, JSON_THROW_ON_ERROR);
    }

    /**
     * Runs one inillucent command and returns its whole result array.
     *
     * The command line's `--output json` contract, unchanged: `ok`, `command`,
     * `columns`, `rows`, `total`, `more`, `changes`, `last_insert_rowid`,
     * `elapsed_ms` and `text` on success, and `ok`, `status`, `message` and
     * sometimes `feature` on failure.
     *
     * @param string $command the verb, such as 'query' or 'describe'
     * @param array<string,mixed> $arguments the named arguments that verb takes
     * @return array<string,mixed>
     * @throws Error when the command could not be run at all
     */
    public function run(string $command, array $arguments = []): array
    {
        $argv = [$this->binary, $command, '--output', 'json', '--db', $this->path];
        if ($this->readOnly) {
            $argv[] = '--readonly';
        }
        if ($this->root !== null) {
            $argv[] = '--root';
            $argv[] = $this->root;
        }
        $stdin = null;
        foreach ($arguments as $name => $value) {
            if ($value === null || $value === false) {
                continue;
            }
            $flag = '--' . str_replace('_', '-', (string) $name);
            if ($value === true) {
                $argv[] = $flag;
                continue;
            }
            // **`params` travels on standard input.** It carries what a command
            // line cannot: a value larger than the argument ceiling, and a
            // binary string, which had no spelling at all before (task-1979,
            // D15 and D17).
            if ($name === 'params' && is_array($value)) {
                $stdin = self::encodeParams(array_values($value));
                $argv[] = '--params-file';
                $argv[] = '-';
                continue;
            }
            $argv[] = $flag;
            // `vector` is a JSON argument and the command line reads it as
            // JSON, so an array is encoded rather than joined.
            $argv[] = is_array($value)
                ? json_encode(array_values($value), JSON_THROW_ON_ERROR)
                : (string) $value;
        }

        $output = Process::run($argv, $stdin);
        $trimmed = trim($output['stdout']);
        // A refusal exits non-zero *and* prints the result object, because JSON
        // was asked for. So the document is what is read, and only an
        // invocation with nothing to parse is a real failure to run.
        if ($trimmed === '' || $trimmed[0] !== '{') {
            throw new Error(
                sprintf(
                    'inillucent %s could not be run (exit %d): %s',
                    $command,
                    $output['code'],
                    trim($output['stderr']) !== '' ? trim($output['stderr']) : $trimmed
                ),
                'internal'
            );
        }

        /** @var array<string,mixed> $result */
        $result = json_decode($trimmed, true, 512, JSON_THROW_ON_ERROR);
        return $result;
    }

    /**
     * Runs a query and returns its rows as associative arrays.
     *
     * @param string $sql the statement
     * @param list<mixed> $params the values for ?1, ?2, ...
     * @param int|null $limit how many rows to hand back
     * @return list<array<string,mixed>>
     * @throws Error when the engine refuses the statement
     */
    public function query(string $sql, array $params = [], ?int $limit = null): array
    {
        $result = $this->run('query', ['sql' => $sql, 'params' => $params ?: null, 'limit' => $limit]);
        $this->refuseFailure($result);
        $names = array_map(static fn (array $column): string => $column['name'], $result['columns']);
        return array_map(
            static fn (array $row): array => array_combine(
                $names,
                array_map(static fn (mixed $value): mixed => self::decodeValue($value), $row)
            ),
            $result['rows']
        );
    }

    /**
     * Turns one cell of a result into the PHP value it stands for.
     *
     * **Bytes went in and could not come back** (task-2066 section 4.1.15). A
     * blob left as the string `x'00ff'`, typed `text` in the column list, so
     * nothing told it apart from a TEXT column holding that text - while the
     * encoder has always sent bytes as `{"blob": "<hex>"}`. The two halves of
     * the same grammar now agree, and a byte string read out of one query binds
     * straight into the next.
     *
     * Anything that is not an envelope is passed through untouched.
     *
     * @param mixed $value one cell, as --output json rendered it
     * @return mixed
     */
    private static function decodeValue(mixed $value): mixed
    {
        if (is_array($value) && array_keys($value) === ['blob'] && is_string($value['blob'])) {
            $bytes = hex2bin($value['blob']);
            return $bytes === false ? $value : $bytes;
        }
        return $value;
    }

    /**
     * Runs one statement for its effect and returns how many rows it changed.
     *
     * @param string $sql the statement
     * @param list<mixed> $params the values for ?1, ?2, ...
     * @throws Error when the engine refuses the statement
     */
    public function exec(string $sql, array $params = []): int
    {
        $result = $this->run('exec', ['sql' => $sql, 'params' => $params ?: null]);
        $this->refuseFailure($result);
        return (int) $result['changes'];
    }

    /**
     * Runs several statements separated by semicolons, as one transaction.
     *
     * Either all of them take effect or none of them do.
     *
     * @param string $sql the statements
     * @throws Error when the engine refuses any of them
     */
    public function batch(string $sql): void
    {
        $this->refuseFailure($this->run('batch', ['sql' => $sql]));
    }

    /**
     * Returns everything about one table: columns, indexes, DDL and row count.
     *
     * @param string $table the table name
     * @return array<string,mixed>
     * @throws Error when there is no such table
     */
    public function describe(string $table): array
    {
        $result = $this->run('describe', ['table' => $table]);
        $this->refuseFailure($result);
        return $result;
    }

    /**
     * Returns the tables and views in the database.
     *
     * @return list<array<string,mixed>>
     * @throws Error when the database cannot be read
     */
    public function tables(): array
    {
        $result = $this->run('tables');
        $this->refuseFailure($result);
        return array_map(
            static fn (array $row): array => ['name' => $row[0], 'type' => $row[1]],
            $result['rows']
        );
    }

    /**
     * Turns a refused result into an exception carrying its status.
     *
     * The status matters more than the message: `unsupported` means the engine
     * has not built the construct and rewording will not help, which is a
     * different thing from a statement being wrong. `Error::isUnsupported()`
     * is how a caller branches on it.
     *
     * @param array<string,mixed> $result what the command line answered
     * @throws Error when the result says it failed
     */
    private function refuseFailure(array $result): void
    {
        if (($result['ok'] ?? false) === true) {
            return;
        }
        throw new Error(
            (string) ($result['message'] ?? 'the command failed'),
            (string) ($result['status'] ?? 'internal'),
            isset($result['feature']) ? (string) $result['feature'] : null
        );
    }
}

File: packages/go/inillucent.go
// Package inillucent drives the inillucent command line from Go and returns
// its results as typed values.
//
// # What this is, and what it is not
//
// This is a wrapper over the `inillucent` binary, not a database driver. Each
// call is a process. That makes it the right tool for the handful of calls a
// build step, a migration script or an agent harness makes, and the wrong tool
// for a loop over a million rows.
//
// It is pure Go with no cgo, which is what lets it build for any target
// without a C toolchain and cross-compile the way the rest of a Go project
// does. A real in-process binding is a worthwhile thing to have and it is a
// different piece of work: it goes through the C ABI in
// drivers/inillucent-driver-capi/include/inillucent_driver.h, with cgo or with
// purego, and drivers/README.md is written to be followed by somebody doing
// exactly that. Writing an unverified cgo binding here and calling it a driver
// would be worse than not having one.
//
// # Getting the binary
//
//	go install github.com/Black-Rainbow-Labs/Inillucent/packages/go/cmd/inillucent-install@latest
//
// installs a small program that downloads the inillucent release for this
// machine, verifies its SHA-256 and puts the four binaries in GOBIN. After that
// this package finds them on PATH.
//
// # Errors that are answers
//
// A refusal comes back as an [Error] carrying a [Status]. StatusUnsupported
// means the engine has not built that construct - it is not a syntax error and
// rewording the statement will not help. Handling the two together throws away
// the distinction this engine's driver was designed around.
package inillucent

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
)

// Status is the class of a failure, as inillucent's driver names them. The same
// thirteen words a C binding reads out of inillucent_error_status.
type Status string

// The statuses. StatusUnsupported is the one worth branching on: it is the
// engine saying it has not built a construct, which is a different thing from
// the statement being wrong.
const (
	StatusUnsupported  Status = "unsupported"
	StatusSyntax       Status = "syntax"
	StatusNotFound     Status = "not_found"
	StatusConstraint   Status = "constraint"
	StatusReadOnly     Status = "readonly"
	StatusBusy         Status = "busy"
	StatusInterrupted  Status = "interrupted"
	StatusCorrupt      Status = "corrupt"
	StatusIO           Status = "io"
	StatusFull         Status = "full"
	StatusTooBig       Status = "too_big"
	StatusInvalidState Status = "invalid_state"
	StatusInternal     Status = "internal"
)

// Column is one result column: the name the statement gave it, and the storage
// class its values turned out to have.
//
// Observed rather than declared, because SQLite's typing is dynamic: a column's
// declared affinity says what it will probably store, and this says what it did.
type Column struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

// Result is what a command produced. It is the command line's `--output json`
// contract, unchanged, so the fields here are the fields a shell script or a
// Python caller sees.
type Result struct {
	OK              bool            `json:"ok"`
	Command         string          `json:"command"`
	Columns         []Column        `json:"columns"`
	Rows            [][]any         `json:"rows"`
	RowCount        int             `json:"row_count"`
	Total           int             `json:"total"`
	More            bool            `json:"more"`
	Changes         int64           `json:"changes"`
	LastInsertRowID int64           `json:"last_insert_rowid"`
	ElapsedMS       float64         `json:"elapsed_ms"`
	Text            string          `json:"text"`
	Status          Status          `json:"status"`
	Message         string          `json:"message"`
	Feature         string          `json:"feature"`
	raw             json.RawMessage `json:"-"`
}

// Field returns a member of the result object that this struct does not name.
//
// Some commands add their own: describe carries `ddl`, `indexes` and
// `row_count_in_table`; stats carries the cache counters. Naming every one of
// them here would make this struct a list that goes stale, so the raw document
// is kept and this reaches into it.
func (result Result) Field(name string, into any) error {
	var members map[string]json.RawMessage
	if err := json.Unmarshal(result.raw, &members); err != nil {
		return err
	}
	held, ok := members[name]
	if !ok {
		return fmt.Errorf("inillucent: %s did not report a %q", result.Command, name)
	}
	return json.Unmarshal(held, into)
}

// Maps returns the rows as maps keyed by column name, which is the shape most
// callers want.
func (result Result) Maps() []map[string]any {
	out := make([]map[string]any, 0, len(result.Rows))
	for _, row := range result.Rows {
		held := make(map[string]any, len(result.Columns))
		for at, column := range result.Columns {
			if at < len(row) {
				held[column.Name] = row[at]
			}
		}
		out = append(out, held)
	}
	return out
}

// Error is a refusal the engine or the command line reported.
type Error struct {
	Status  Status
	Message string
	Feature string
	Command string
}

// Error makes this an error, naming the class as well as the message so a log
// line says which kind of failure it was.
func (e *Error) Error() string {
	if e.Feature != "" {
		return fmt.Sprintf("inillucent %s [%s]: %s (not built: %s)", e.Command, e.Status, e.Message, e.Feature)
	}
	return fmt.Sprintf("inillucent %s [%s]: %s", e.Command, e.Status, e.Message)
}

// Unsupported reports whether an error is the engine saying it has not built
// the construct, rather than the statement being wrong.
//
//	if inillucent.Unsupported(err) {
//	        // do not offer that feature; rewording will not help
//	}
func Unsupported(err error) bool {
	held, ok := err.(*Error)
	return ok && held.Status == StatusUnsupported
}

// DB is a database, addressed by its path. It is not a connection and holds
// nothing open: every call is a process, so a DB is safe to share across
// goroutines and safe to keep for the life of a program.
type DB struct {
	// Path is the database file. Empty means an in-memory database, which is
	// discarded when the process running the command ends - so it is only
	// useful for a single Batch.
	Path string
	// Binary is the inillucent program to run. Empty means "inillucent", found
	// on PATH.
	Binary string
	// ReadOnly refuses every statement that changes something, classified by
	// the binder rather than by scanning the text.
	ReadOnly bool
	// Root, when set, refuses every path outside that directory.
	Root string
}

// Open returns a DB for a database file. It does no I/O: the file is opened by
// whichever command runs first, and a path that does not exist is created by
// the first command that writes.
func Open(path string) *DB {
	return &DB{Path: path}
}

// binary returns the program to run.
func (db *DB) binary() string {
	if db.Binary != "" {
		return db.Binary
	}
	if named := os.Getenv("INILLUCENT_BIN"); named != "" {
		return named
	}
	return "inillucent"
}

// Run executes one inillucent command and returns its result.
//
// The arguments are the named ones that verb takes: Run(ctx, "describe",
// Args{"table": "people"}). A refusal is returned as an *Error rather than in
// the Result, so the ordinary `if err != nil` shape works; the Result is
// returned too, for the callers that want the timing or the raw text.
func (db *DB) Run(ctx context.Context, command string, arguments Args) (Result, error) {
	argv := []string{command, "--output", "json"}
	if db.Path != "" {
		argv = append(argv, "--db", db.Path)
	}
	if db.ReadOnly {
		argv = append(argv, "--readonly")
	}
	if db.Root != "" {
		argv = append(argv, "--root", db.Root)
	}
	argv = append(argv, arguments.flags()...)

	run := exec.CommandContext(ctx, db.binary(), argv...)
	output, runErr := run.Output()
	// A refusal exits non-zero *and* prints the result object, because the
	// caller asked for JSON. So the exit status is not the thing to read - the
	// document is - and only an invocation with nothing to parse is a real
	// failure to run.
	trimmed := strings.TrimSpace(string(output))
	if !strings.HasPrefix(trimmed, "{") {
		if runErr != nil {
			var exited *exec.ExitError
			if ok := asExitError(runErr, &exited); ok && len(exited.Stderr) > 0 {
				return Result{}, fmt.Errorf("inillucent %s: %s", command, strings.TrimSpace(string(exited.Stderr)))
			}
			return Result{}, fmt.Errorf("inillucent %s: %w", command, runErr)
		}
		return Result{}, fmt.Errorf("inillucent %s produced no result object", command)
	}

	var result Result
	if err := json.Unmarshal([]byte(trimmed), &result); err != nil {
		return Result{}, fmt.Errorf("inillucent %s: could not read its result: %w", command, err)
	}
	result.raw = json.RawMessage(trimmed)
	if !result.OK {
		return result, &Error{
			Status:  result.Status,
			Message: result.Message,
			Feature: result.Feature,
			Command: command,
		}
	}
	return result, nil
}

// asExitError is errors.As, spelled out so this package imports nothing beyond
// the standard library's smallest useful set.
func asExitError(err error, into **exec.ExitError) bool {
	held, ok := err.(*exec.ExitError)
	if ok {
		*into = held
	}
	return ok
}

// Args are a command's named arguments. A []any is serialised as JSON, which is
// what `params` and `vector` expect; a bool becomes a bare flag when true and is
// left out when false.
type Args map[string]any

// flags renders the arguments as command-line words.
func (args Args) flags() []string {
	out := make([]string, 0, len(args)*2)
	for name, value := range args {
		flag := "--" + strings.ReplaceAll(name, "_", "-")
		switch held := value.(type) {
		case nil:
			continue
		case bool:
			if held {
				out = append(out, flag)
			}
		case string:
			out = append(out, flag, held)
		case int:
			out = append(out, flag, strconv.Itoa(held))
		case int64:
			out = append(out, flag, strconv.FormatInt(held, 10))
		case float64:
			out = append(out, flag, strconv.FormatFloat(held, 'g', -1, 64))
		default:
			// Everything else - a slice of values for `params`, a vector - is a
			// JSON argument, and the command line reads it as JSON.
			encoded, err := json.Marshal(held)
			if err != nil {
				continue
			}
			out = append(out, flag, string(encoded))
		}
	}
	return out
}

// Query runs a statement that returns rows.
//
//	rows, err := db.Query(ctx, "SELECT id, name FROM people WHERE id > ?1", 3)
func (db *DB) Query(ctx context.Context, sql string, params ...any) (Result, error) {
	arguments := Args{"sql": sql}
	if len(params) > 0 {
		arguments["params"] = params
	}
	return db.Run(ctx, "query", arguments)
}

// Exec runs one statement for its effect and returns how many rows it changed.
func (db *DB) Exec(ctx context.Context, sql string, params ...any) (int64, error) {
	arguments := Args{"sql": sql}
	if len(params) > 0 {
		arguments["params"] = params
	}
	result, err := db.Run(ctx, "exec", arguments)
	return result.Changes, err
}

// Batch runs several statements separated by semicolons, as one transaction.
// Either all of them take effect or none of them do.
func (db *DB) Batch(ctx context.Context, sql string) error {
	_, err := db.Run(ctx, "batch", Args{"sql": sql})
	return err
}

// Describe returns everything about one table in a single call: its columns,
// its indexes, its DDL and how many rows it holds.
func (db *DB) Describe(ctx context.Context, table string) (Result, error) {
	return db.Run(ctx, "describe", Args{"table": table})
}

// Tables lists the tables and views.
func (db *DB) Tables(ctx context.Context) (Result, error) {
	return db.Run(ctx, "tables", nil)
}

// Search runs a full-text query over an FTS5 or inillucent_search table.
func (db *DB) Search(ctx context.Context, table, query string, k int) (Result, error) {
	arguments := Args{"table": table, "query": query}
	if k > 0 {
		arguments["k"] = k
	}
	return db.Run(ctx, "search", arguments)
}

// Version returns what the installed inillucent calls itself, which is also the
// cheapest way to check that the binary is on PATH and runnable.
func Version(ctx context.Context) (string, error) {
	db := &DB{}
	result, err := db.Run(ctx, "version", nil)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(result.Text), nil
}
